### PR TITLE
Refactor the payment status

### LINKED ISSUE #6998: Refactor payment status [after data store work is merged]
Thile working on [Feature: Data Store Migration](https://github.com/woocommerce/woocommerce-blocks/pull/6612#top), I realised thy payment status was implemented in quite a confusing way. On trunk, pre data store work, we had a `payment-method` state using a reducer, and a `payment-method-data-context` context that used some of that state, added to it, and exposed, well, more state, but as context. We tried to simplify this by combining both of those pieces of state into one and storing that in the data store. For the most part, this was successful, apart from the status. 

What we have now (post data store work), is a bit counterintuitive. We have a `currentStatus` state variable, with multiple boolean sub properties for each possible status. Something like this:

```js
{
    currentStatus: {
		isPristine: true,
		isStarted: false,
		isProcessing: false,
		isFinished: false,
		hasError: false,
		hasFailed: false,
		isSuccessful: false,
		isDoingExpressPayment: false,
	},
     ...
```

What would be more useful is to work out what the minimum state we need to store, and convert some of these properties to selectors, like we do in the checkout data store.

We also want to make sure that when setting the currentStatus to `pristine`, we are resetting all the necessary values to the default state.

### PR DESCRIPTION
**Pull request by alexflorisca**
The payment status in the payment data store is composed of an object with a bunch of boolean properties that don't always make sense. For example `isProcessing` can be true at the the same time as `isPristine`. See #6998 for more details.

This PR refactors the payment status to have one string value with the status. The various boolean helpers are now selectors. Things to note:

- Replaced the boolean values inside the `currentStatus` object with selectors for querying the payment status
- We need to keep the API we expose to third parties consistent, so the `usePaymentMethodInterface` hook exposes the `paymentStatus` as it was before. 
- Removed the `__internalSetCurrentStatus` action because it is marked as internal, hasn't been advertised to third party devs and hasn't been around that long (a month or two since we merged the data store work);
- Deprecated but kept the `getCurrentStatus` selector as it is documented and available to third parties, even though this is fairly new so I doubt anyone has had a chance to implement it. 

Fixes #6998 

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [x] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Load the checkout block and open up the payments store in Redux dev tools
2. You should see 2 calls to `SET_PAYMENT_PRISTINE`. One comes after we set the default payment method and the other because the checkout is idle and we haven’t completed the payment yet
3. Click on an express payment method. You should see the `SET_PAYMENT_STARTED` action fired and the `status` change to `started`
4. Click out of the express payment modal. You should see the `SET_PAYMENT_PRISTINE` action fired and status changed to `pristine`
5. Open up [[this file](https://github.com/woocommerce/woocommerce-blocks/blob/d295a71606d546e6f8be66302380b6a149ab4001/assets/js/data/checkout/utils.ts#L168)](https://github.com/woocommerce/woocommerce-blocks/blob/d295a71606d546e6f8be66302380b6a149ab4001/assets/js/data/checkout/utils.ts#L168) and add `dispatch.__internalSetHasError( true );` (and maybe a console log) on line 167.
6. Go through the checkout flow with a stripe CC and check the `SET_PAYMENT_PRISTINE` action is fired and status changed to `pristine` and the console log shows. This is a really unlikely (impossible maybe?) scenario that resets the payment status to pristine, given an error with the checkout, **after** the payment has been successful.
7. Remove the code above.
8. Go through the checkout process once again. Check the `SET_PAYMENT_PROCESSING` action is fired and the status changed to `processing` and the `SET_PAYMENT_METHOD_DATA` and `SET_PAYMENT_SUCCESS` actions are fired and the status and paymentMethodData keys are updated in the store
9. The `SET_PAYMENT_FAILED` and `SET_PAYMENT_ERROR` are difficult to test as it requires modifying code in the stripe plugin. They are also not accurate representations of the state which I have explained in #7667. For now I think it’s ok to check the code and make sure these set the correct value on the status.

* [x] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Refactored the payment store to provide clearer payment status via new wp/data selectors

### CODE REVIEW

File: assets/js/data/payment/selectors.ts [hunk 1]
```diff
@@ -73,8 +100,21 @@ export const expressPaymentMethodsInitialized = (
 	return state.expressPaymentMethodsInitialized;
 };
 
+/**
+ * @deprecated - use these selectors instead: isPaymentPristine, isPaymentStarted, isPaymentProcessing,
```
Comment (opr) [inline assets/js/data/payment/selectors.ts:105]: Do you know about the `deprecated` function from [`@wordpress/deprecated`](https://github.com/WordPress/gutenberg/blob/trunk/packages/deprecated/README.md) ? I think we should use that here.

File: docs/third-party-developers/extensibility/data-store/payment.md [hunk 2]
```diff
@@ -27,6 +27,136 @@ with. We do not encourage extensions to dispatch actions onto this data store ye
 
 ## Selectors
 
+### isPaymentPrestine
```
Comment (opr) [inline docs/third-party-developers/extensibility/data-store/payment.md]:
```suggestion
### isPaymentPristine
```

File: docs/third-party-developers/extensibility/data-store/payment.md [hunk 3]
```diff
@@ -27,6 +27,136 @@ with. We do not encourage extensions to dispatch actions onto this data store ye
...
+
+Queries if the status is `pristine` or not.
```
Comment (opr) [inline docs/third-party-developers/extensibility/data-store/payment.md]: Nitpicky, but could we reword to: "Returns whether the status is `pristine`.

Feel free to ignore I just think it sounds a little better without the "or not". This applies to the rest of the doc too.

File: docs/third-party-developers/extensibility/data-store/payment.md [hunk 4]
```diff
@@ -27,6 +27,136 @@ with. We do not encourage extensions to dispatch actions onto this data store ye
...
+
+```js
+const store = select( 'wc/store/payment' );
+const isPaymentFinished = store.isPaymentFinished();
+```
+
+### getCurrentStatus (deprecated)
+
+Returns an object with booleans representing the payment status.
+_**This selector is deprecated and will be removed in a future release. Please use the 7 selectors above instead to query the payment status**_
```
Comment (opr) [inline docs/third-party-developers/extensibility/data-store/payment.md]: Maybe remove the 7 from here, and just have "please use the selectors above"?

**Review by opr — CHANGES_REQUESTED** [review]
Hey @alexflorisca nice work! Just a couple of things:

In the testing instructions:

> You should see 2 calls to SET_PAYMENT_PRISTINE. One comes after we set the default payment method and the other because the checkout is idle and we haven’t completed the payment yet

I only see this once. Just want to double check that this doesn't indicate something being broken?

> Go through the checkout flow with a stripe CC and check the SET_PAYMENT_PRISTINE action is fired and status changed to pristine and the console log shows. This is a really unlikely (impossible maybe?) scenario that resets the payment status to pristine, given an error with the checkout, after the payment has been successful.

I see this, and I see the error briefly before being redirected to the success page, is this the expected behaviour?

File: docs/third-party-developers/extensibility/data-store/payment.md (hunk 2 above)
Comment (alexflorisca) [reply docs/third-party-developers/extensibility/data-store/payment.md]: This was Copilot haha

File: assets/js/data/payment/selectors.ts (hunk 1 above)
Comment (alexflorisca) [reply assets/js/data/payment/selectors.ts:104]: I did not, thanks for pointing it out, I've added it in now

Comment (alexflorisca) [reply assets/js/data/payment/selectors.ts:105]: I didn't thanks for pointing it out, I've changed it now

**Comment by alexflorisca** [comment]
Thanks for the review @opr , I've added your suggestions in now. 

> I only see this once. Just want to double check that this doesn't indicate something being broken?

Yeah it's fine, it's probably because you have some saved payment methods, in which case it will skip running one of them.

> I see this, and I see the error briefly before being redirected to the success page, is this the expected behaviour?

Yep that's expected. I forgot to say in the testing instructions you can comment out the `window.location...` line in the `checkout-processor.js` to see the actions more clearly.

File: assets/js/data/payment/selectors.ts [hunk 5]
```diff
@@ -105,6 +106,13 @@ export const expressPaymentMethodsInitialized = (
  * isPaymentFinished, hasPaymentError, isPaymentSuccess, isPaymentFailed
  */
 export const getCurrentStatus = ( state: PaymentMethodDataState ) => {
+	deprecated( 'getCurrentStatus', {
```
Comment (opr) [inline assets/js/data/payment/selectors.ts:109]: Please could you add a link to this PR in the deprecated message?

Comment (opr) [reply assets/js/data/payment/selectors.ts:109]: Like here: https://github.com/woocommerce/woocommerce-blocks/blob/4fcb6b28d69903b00dd296dec08db32a8d5ee67a/packages/checkout/filter-registry/index.ts#L66

Comment (alexflorisca) [reply assets/js/data/payment/selectors.ts:109]: Done :)

**Review by opr — APPROVED** [review]
Thanks for making all these changes, looks great now! 🥳

**Pull request merged**